src/model_theory/rings/language.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2022 Joseph Hua. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Hua
+-/
+import model_theory.terms_and_formulas
+
+/-!
+# The language of rings
+
+## Main Definitions
+* A `first_order.language.ring.L` defines the language of rings,
+  which consists of `0`, `1`, `-`, `+`, `*`.
+-/
+universes u v u'
+
+namespace first_order
+namespace language
+namespace ring
+
+/-- The constant symbols in `first_order.language.ring.L` -/
+inductive consts : Type*
+| zero : consts
+| one : consts
+
+/-- The unary function symbols in `first_order.language.ring.L` -/
+inductive unaries : Type*
+| neg : unaries
+
+/-- The binary function symbols in `first_order.language.ring.L` -/
+inductive binaries : Type*
+| add : binaries
+| mul : binaries
+
+/-- All function symbols in `first_order.language.ring.L` -/
+def functions : ℕ → Type*
+| 0 := consts
+| 1 := unaries
+| 2 := binaries
+| (n + 3) := pempty
+
+instance : inhabited consts := ⟨ consts.zero ⟩
+instance : inhabited unaries := ⟨ unaries.neg ⟩
+instance : inhabited binaries := ⟨ binaries.add ⟩
+instance : inhabited (functions 0) := ⟨ consts.zero ⟩
+
+/-- The language of rings -/
+def L : language.{u v} :=
+{ functions := functions,
+  relations := λ n, pempty }
+
+variable {α : Type u'}
+
+@[simp] instance : has_zero (L.term α) := ⟨ @func L _ 0 consts.zero fin_zero_elim ⟩
+
+@[simp] instance : has_one (L.term α) := ⟨ @func L _ 0 consts.one fin_zero_elim ⟩
+
+@[simp] instance : has_neg (L.term α) := ⟨ λ x, @func L _ 1 unaries.neg ![x] ⟩
+
+@[simp] instance : has_add (L.term α) :=
+⟨ λ x y, @func L _ 2 binaries.add ![x, y] ⟩
+
+@[simp] instance : has_mul (L.term α) :=
+⟨ λ x y, @func L _ 2 binaries.mul (λ i, match i with | ⟨0, _⟩ := x | ⟨n+1, _⟩ := y end) ⟩
```

Comment:
```suggestion
⟨ λ x y, @func L _ 2 binaries.mul ![x, y] ⟩
```

Reply:
Is there a reason you marked this as resolved?

Reply:
Oh I got confused and missed this one out.